### PR TITLE
@0xproject/deployer improvements

### PR DESCRIPTION
**Pull request by LogvinovLeon**
## Motivation and Context

We're preparing deployer for a public release.

## How Has This Been Tested?

Deployer tests

## Types of changes

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [ ] New feature (non-breaking change which adds functionality)
*   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [x] Change requires a change to the documentation.
*   [x] Added tests to cover my changes.
*   [x] Added new entries to the relevant CHANGELOGs.
*   [x] Updated the new versions of the changed packages in the relevant CHANGELOGs.
*   [x] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [x] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/deployer/src/compiler.ts [hunk 1]
```diff
@@ -23,17 +32,13 @@ import {
...
-const DEPENDENCY_PATH_REGEX = /"([^"]+)"/; // Source: https://github.com/BlockChainCompany/soljitsu/blob/master/lib/shared.js
 
 export class Compiler {
     private _contractsDir: string;
     private _networkId: number;
-    private _optimizerEnabled: number;
+    private _optimizerEnabled: boolean;
     private _artifactsDir: string;
-    private _contractSources?: ContractSources;
+    private _contractSources!: ContractSources;
```
Comment (fabioberger) [inline packages/deployer/src/compiler.ts]: Can you add a comment explaining the need for `!` here?

File: packages/deployer/src/compiler.ts [hunk 2]
```diff
@@ -206,28 +159,31 @@ export class Compiler {
...
-                const normalizedErrMsg = Compiler._getNormalizedErrMsg(errMsg);
+                const normalizedErrMsg = getNormalizedErrMsg(errMsg);
                 this._solcErrors.add(normalizedErrMsg);
             });
         }
         const contractName = path.basename(fileName, constants.SOLIDITY_FILE_EXTENSION);
         const contractIdentifier = `${fileName}:${contractName}`;
+        if (_.isUndefined(compiled.contracts[contractIdentifier])) {
+            throw new Error(
+                `Contract ${contractName} not found in ${fileName}. Please make sure your contract has the same name as a file`,
```
Comment (fabioberger) [inline packages/deployer/src/compiler.ts]: `as a file` -> `as it's file name`

File: packages/deployer/src/compiler.ts [hunk 3]
```diff
@@ -273,79 +230,42 @@ export class Compiler {
...
-        } catch (err) {
-            logUtils.log(`Artifact for ${fileName} does not exist`);
-            return undefined;
+    private _getSourceTreeHash(fileName: string, sourceHash: Buffer, dependencies: string[]): Buffer {
+        if (dependencies.length === 0) {
+            return sourceHash;
+        } else {
+            const dependencySourceTreeHashes = _.map(dependencies, dependency => {
+                const source = this._contractSources[dependency];
+                this._setContractSpecificSourceData(source, dependency);
```
Comment (fabioberger) [inline packages/deployer/src/compiler.ts]: Hm, this is a code smell. Why is there a `set` op. inside a `getter` function? Isn't this a side-effect?

File: packages/deployer/src/utils/compiler.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import * as solc from 'solc';
+
+import { constants } from './constants';
+import { fsWrapper } from './fs_wrapper';
+import { ContractArtifact, ContractSources } from './types';
+
+/**
+ * Gets contract data on network or returns if an artifact does not exist.
+ * @param fileName Name of contract file.
+ * @return Contract data on network or undefined.
```
Comment (fabioberger) [inline packages/deployer/src/utils/compiler.ts]: Missing second param comment.

File: packages/deployer/src/utils/compiler.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        return contractArtifact;
+    } catch (err) {
+        logUtils.log(`Artifact for ${fileName} does not exist`);
+        return undefined;
+    }
+}
+
+/**
+ * Creates the artifacts directory if it does not already exist.
+ */
```
Comment (fabioberger) [inline packages/deployer/src/utils/compiler.ts]: Missing param comment.

File: packages/deployer/src/utils/compiler.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        }
+    });
+    return dependencies;
+}
+
+/**
+ * Callback to resolve dependencies with `solc.compile`.
+ * Throws error if contractSources not yet initialized.
+ * @param  importPath Path to an imported dependency.
+ * @return Import contents object containing source code of dependency.
```
Comment (fabioberger) [inline packages/deployer/src/utils/compiler.ts]: Missing param comment.

File: packages/deployer/src/utils/types.ts [hunk 7]
```diff
@@ -84,9 +84,9 @@ export interface ContractSourceData {
 
 export interface ContractSpecificSourceData {
     dependencies: string[];
-    solcVersion: string;
+    solcVersionRange: string;
     sourceHash: Buffer;
-    sourceTreeHashIfExists?: Buffer;
+    sourceTreeHash: Buffer;
```
Comment (fabioberger) [inline packages/deployer/src/utils/types.ts]: Mind addressing the TODO below? We now have types in a separate package.

File: packages/deployer/src/compiler.ts (hunk 3 above)
Comment (LogvinovLeon) [reply packages/deployer/src/compiler.ts]: It is, but this is how it's currently designed. I'm planning to redesign th interface in the next PR. This one already makes it more clear.

**Pull request merged**